### ISSUE TITLE
Cluster DNS example - connection timeout to backend

### ISSUE THREAD

**Issue by marekbiskup** [GoogleCloudPlatform/kubernetes#11207]
This is about examples/cluster-dns/.

I'm trying to write an automated test and the code is more or less the following:

```
runKubectl("create", "-f", backendRcYaml, getNsCmdFlag(ns))  // these pods print "Hello world"
runKubectl("create", "-f", backendSvcYaml, getNsCmdFlag(ns))
waitForRCPodsRunning(c, ns.Name, backendRcName)   // polls pod status on the API server
waitForService(c, ns.Name, backendSvcName, true, poll, serviceStartTimeout)  // polls the API server
err = podsResponding(c, ns.Name, backendPodName, false, pods) // confirms both pods answer ''hello world" via the proxy (from the client)
// I added the code below to make sure the DNS already knows about the service
// this python is executed in the container of the first pod.
// I assume this queries the DNS (not sure though).
[python]
   import socket
   try:
	socket.gethostbyname('%s')
	print 'ok'
   except:
	print 'err'`
[/python]

newKubectlCommand("create", "-f", "-", getNsCmdFlag(ns)).withStdinData(updatedPodYaml).exec()   // execute the frontend pod - should query the backend and print the result ("hello world")
_, err := lookForStringInLog(ns.Name, frontendPodName, frontendPodContainerName, podOutput, podStartTimeout)  // look for "Hello world"
```

The frontend pod just tries to connect the backend via a dns name and then prints the result. In most cases it works great, but in 1/20 cases instead of the result, the frontend pod gets a timeout.

```
Failed to find "Hello World!", last result: "10.0.174.97
Send request to: http://dns-backend.e2e-tests-dnsexample0-dgv5v.cluster.local:8000
Traceback (most recent call last):
  File "client.py", line 30, in <module>
    Main()
  File "client.py", line 26, in Main
    GetServerResponse(args.address)
  File "client.py", line 16, in GetServerResponse
    response = requests.get(address)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 69, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 465, in request
     resp = self.send(prep, **send_kwargs)
     File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', error(110, 'Connection timed out'))"
```

I believe there is something wrong with kubernetes, not with the application.
The PR for the DNS test is  #11050

cc @ddysher @thockin

**Comment by marekbiskup** [comment]
Note that even though this occurs quite infrequently, it is quite easy to reproduce. Usually it takes 30-90 minutes to reproduce.

**Comment by marekbiskup** [comment]
cc @mwielgus

**Comment by marekbiskup** [comment]
@mwielgus suggests this is a problem with delayed endpoint creation for a service. Basically, when I connect to the service, it has no endpoints at the moment and that's why connection times out.

The solution is to wait for service endpoints before creating the frontend pod.

**Comment by thockin** [comment]
If you have already verified that DNS answers, the endpoints exist.  That
said, we know DNS has a bug wherein the entries are briefly deleted and
recreated, but that should only be every 30 minutes.

Can we distill this failure down to a simple shell script?

**Comment by brendandburns** [comment]
Also, Pod running is no guarantee of Pod serving, you should make sure that you have readiness checks enabled on your Pods.

**Comment by thockin** [comment]
I'm running a DNS lookup every 0.2 seconds for the last hour and saw 0
failures to resolve in my cluster.

**Comment by marekbiskup** [comment]
Dns works because it is confirmed by the python code. Also the pods answer because I call `podsResponding()`
Finally, the stack trace shows that the error comes from GetServerResponse and the code is like this:

```
import argparse
import requests
import socket

from urlparse import urlparse

def CheckServiceAddress(address):
  hostname = urlparse(address).hostname
  service_address = socket.gethostbyname(hostname)
  print service_address


def GetServerResponse(address):
  print 'Send request to:', address
  response = requests.get(address)
  print response
  print response.content


def Main():
  parser = argparse.ArgumentParser()
  parser.add_argument('address')
  args = parser.parse_args()
  CheckServiceAddress(args.address)
  GetServerResponse(args.address)


if __name__ == "__main__":
  Main()
```

Clearly, pods are responding, DNS is responding, but it is not possible to connect to the IP address of a pod. 

I'm going to add a loop that waits for endpoints before starting the frontend pod (with the code shown above, in this comment). Then I'll run the test for a night.

**Comment by marekbiskup** [comment]
I added querying the service via the proxy (from the client) before trying to query the service via DNS (from another pod) and the issue has gone.
I'm not 100% sure IP creation delay was the cause. Maybe just an additional delay, alone, helped to cover another issue. (I use wait.Poll, which adds 5s delay before makes a check.)
